elastic/kibana
# [TSVB] Runs the migration correctly for the drop last bucket

## Summary

Fixes https://github.com/elastic/kibana/issues/121684

### The problem
On 7.14 we changed the default behavior of the `Drop last bucket` setting from Yes to No. We also created another PR that adds a migration script (for < 7.14.0 SOs) #110782. What this migration does:
- Checks if the global `drop_last_bucket` setting is set. If no, it adds the property with 1 (Drop last bucket ? Yes)
- Checks if the chart uses the override dataview setting, and changes the `series_drop_last_bucket ` to 1

This works fine for all charts / dashboards that migrate from a version prior to 7.14.0 (and for the metricbeats new installations)

In 7.15, beats changed their visualizations SOs (for example here https://github.com/elastic/beats/blob/master/metricbeat/module/system/_meta/kibana/7/visualization/83e12df0-1b91-11e7-bec4-a5e9ec5cab8b-ecs.json) and the new migration version became v7.14.0. This means that the above migration doesn't run for the newly created metricbeats dashboards (7.15++)

### The fix
I created another migration (7.17). This is not the same migration that runs for 7.14.0 because we can't use it unfortunately if the override index pattern setting is used. This migration checks if the global `drop_last_bucket` setting is set. If no, it adds the property with 1 (Drop last bucket ? Yes)

This is safe as the `drop_last_bucket` is saved with the SO regardless the value from 7.14++. This means:
- For a prior to 7.14 viz, the first migration will run and add the value if it doesn't exist.
- For viz created for 7.14 and above, this value is saved in the SO and the second migration will run but won't affect the SO as the property exists.

Unfortunately this is not safe for the visualizations that are using the override index pattern setting and have been created for 7.14 +. The reason is that the if the default value of `series_drop_last_bucket ` (default is No), this is not saved in the SO. What it means if I run the migration:
-  For a prior to 7.14 viz, the first migration will run and add the value if it doesn't exist. Then the second migration will run but the property has already be set so nothing will change to the SO. Everything good so far.
- BUT for a viz that was created on 7.14++ that uses the override index pattern and uses the default value (drop last bucket No), if I run the migration that created on 7.14.0, it will set this value to Yes but this is not what the user wants. The user used the default value which is No.
The above cases can't be solved by us unfortunately. I think that they should be changed on the beats side. As far as I can see they are 4 beats charts that use the 'override_index_pattern : 1'  property https://github.com/elastic/beats/search?p=1&q=override_index_pattern. I think that we should update them with adding `series_drop_last_bucket : 1`

In order this to never happen again, this PR except from adding the migration, also adds the default values for the `override_index_pattern ` and `series_drop_last_bucket` for a newly created TSVB chart. With this change, the default values will always be stored in the SO.

### How to test

1. Install the metricbeat dashboards and check the gauges visualizations. The `drop_last_bucket` setting should be set to Yes
v8.1 newly installed metricbeat dashboard
![image](https://user-images.githubusercontent.com/17003240/146926562-7c8ada14-6351-4472-9f7e-26d53c95da56.png)

and the dashboard
![image](https://user-images.githubusercontent.com/17003240/146926800-800c911d-cb7a-4de3-a2ce-2a35bc708ee2.png)


1. Create a prior to 7.14 viz with different values of the drop last bucket and migrate it locally. Check the viz that they are working as expected. Test by value and by ref charts.
2. Do the same for a 7.14++ viz. Test by value and by ref charts.

### Checklist
